### PR TITLE
Dynamic caching logic

### PR DESCRIPTION
**Pull request by mckn**
This PR adds dynamic caching functionality that will apply caching logic based in the query and time range.

Closes: #125

Skärmavbild 2020-09-16 kl  22 41 06

### CODE REVIEW

**Review by hugohaggmark — APPROVED** [review]
Frontend looks good 🌟

**Comment by mckn** [comment]
I am waiting on a couple of answers from the people at MS regarding some of the edge cases. Will add those as tests when I get a reply.

File: pkg/azuredx/cache_test.go [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+				},
+			},
+		},
+		{
+			name: "should adjust settings based on query bin size 1h",
+			configuration: dataSourceData{
+				CacheMaxAge:    "10s",
+				DynamicCaching: true,
+			},
+			query: `SampleData\n` +
```
Comment (kylebrandt) [inline pkg/azuredx/cache_test.go:71]: I think the diff ends up funny when this fails with using `+`, so better to use three backticks instead

File: pkg/azuredx/cache.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+	if diff > time.Hour {
+		return time.Minute
+	}
+
+	return cacheMinAge
+}
+
+// binSizeRE finds the bin size of a query by looking at the summarize statement
+// e.g. summarize avg(Column), count(Column) by bin(TimeColumn, 1d)
+var binSizeRE = regexp.MustCompile(`summarize [\w\$\(\)\.,\s]+ by bin\([\w\$\(\)\.]+, ` +
```
Comment (kylebrandt) [inline pkg/azuredx/cache.go:93]: Worry about change query behavior based on a regex. But if we have to do it, I think we should just detect the bin.

**Comment by kminehart** [comment]
out of curiosity, we have a couple of enterprise datasources that have caching built in; usually it's just an in-memory map. Is this a similar thing or some cache options provided by Azure?

**Comment by mckn** [comment]
> out of curiosity, we have a couple of enterprise datasources that have caching built in; usually it's just an in-memory map. Is this a similar thing or some cache options provided by Azure?

This is basically just setting cache headers that are being sent to Azure that will trigger them to cache queries. But what we have added is the possibility to either set these values manually or enabling our datasource plugin to detect what caching options would be optimal on a per-query basis.
